fs/zbd_zenfs.cc

```diff
@@ -631,43 +651,108 @@ Zone *ZonedBlockDevice::AllocateZone(Env::WriteLifeTimeHint file_lifetime, bool
   LatencyHistGuard guard(&io_alloc_latency_reporter_);
   io_alloc_qps_reporter_.AddCount(1);
 
-  io_zones_mtx.lock();
+  auto t0 = std::chrono::system_clock::now();
+
+  // For general data, we need both two locks, so the general data thread
+  // can give up CPU to WAL thread.
+  if (!is_wal) {
+    io_zones_mtx.lock();
+    AtomicWriter() << std::this_thread::get_id() << " io_zones_mtx.lock()\n";
+  } else {
+    wal_zone_allocating_++;
+  }
 
   /* Make sure we are below the zone open limit */
   {
     std::unique_lock<std::mutex> lk(zone_resources_mtx_);
-    zone_resources_.wait(lk, [this] {
-      if (open_io_zones_.load() < max_nr_open_io_zones_) return true;
+    zone_resources_.wait(lk, [this, is_wal] {
+      if (!is_wal) {
+        return open_io_zones_.load() < max_nr_open_io_zones_ - 3;
+      } else {
+        return open_io_zones_.load() < max_nr_open_io_zones_;
+      }
       return false;
     });
   }
 
+  wal_zones_mtx.lock();
+  AtomicWriter() << std::this_thread::get_id() << " wal_zones_mtx.lock()\n";
+  if (is_wal) wal_zone_allocating_--;
+
   LatencyHistGuard guard_actual(&io_alloc_actual_latency_reporter_);
 
+  uint64_t reset_max = 0;
+  uint64_t finish_max = 0;
+  auto t1 = std::chrono::system_clock::now();
+
+  AtomicWriter() << std::this_thread::get_id()
+                 << " \tacquire lock: " << TimeDiff(t0, t1) << "\n";
+
   /* Reset any unused zones and finish used zones under capacity treshold*/
-  for (const auto z : io_zones) {
+  for (int i = 0; i < io_zones.size(); i++) {
+    // unlock wal mutex and give wal thread a chance
+    if (!is_wal && wal_zone_allocating_.load() > 0) {
+      wal_zones_mtx.unlock();
+      AtomicWriter() << std::this_thread::get_id() << " " << i
+                     << " wal_zones_mtx.unlock()"
+                     << "\n";
+      while (wal_zone_allocating_.load() > 0) {
```

Comment:
Why not using a semaphore for `wal_zone_allocating_`?

Reply:
We may have multiple pending WAL allocations, so if there any existing WAL writes are pending we want all background writes to give up running.